doc/manuals/sprokit/embedded_pipeline.rst

```diff
@@ -0,0 +1,327 @@
+always that realistic for more complicated pipelines because there is
+usually some latency so the output will not be available until several
+inputs have been supplied.::
+
+       for ( int i = 0; i < 10; ++i)
+       {
+         // Create dataset for input
+   11)   auto ds = kwiver::adapter::adapter_data_set::create();
+
+         // Add value to be pushed to the named port
+   12)   ds.add_value( "counter", i );
+
+        // Data values need to be supplied to all connected ports
+        // (based on the previous pipeline definition)
+   13)  ds.add_value( "port2", i );
+   14)  ds.add_value( "port3", i );
+   15)  ep.send( ds ); // push into pipeline
+
+        // Get output from pipeline
+   16)  auto rds = ep.receive();
+
+        // get value from the output adapter
+   17)  int val = get_port_data<int>( "out_num" );
+      } // end for
+
+Line 11 creates a new adapter data set object. An adapter_data_set
+contains all inputs to the pipeline. They are collected in this object
+so they can be presented to the pipeline at the same time.
+
+Lines 12 - 14 add individual data values to the pipeline input object
+(adapter_data_set). The string specified in the call must match the
+port name that was used to connect to the input_adapter. The value
+specified will be supplied to that port.
+
+Line 15 sends the set of input data to the input adapter process. An
+error will be thrown if there is a port connected to that process
+which does not have an associated data element. An error will also be
+thrown if there is a element with a name that is not connected to the
+input process.
+
+Line 17 retrieves a set of output values from the pipeline. There will
+be a value for each port that is connected to the output_process.
+
+
+Pipeline Inputs and Outputs
+---------------------------
+
+In order to adapt a pipeline to running in an embedded manner, the
+inputs that are supplied by the program are passed to the
+*input_adapter* process and the outputs from the pipeline are passed to
+the *output_adapter*. The pipeline definition must specify the
+connections from/to these processes.
+
+Sets of input data elements are passed to the pipeline using a
+`adapter_data_set` object. This class defines a names set of data
+items where the name corresponds to the port name, as specified in the
+pipeline definition. The type of the data element must be compatible
+with what is expected on the port by the receiving process. The
+*output_adapter* returns the named data elements in the same way.
+
+..  doxygenclass:: kwiver::adapter::adapter_data_set
+    :project: kwiver
+    :members:
+
+Polling the interface queues
+----------------------------
+
+The above example code uses `send()` and `receive()` in a loop to
+supply data to the pipeline and retrieve the output. While this is a
+direct approach, it will not work if there is any latency in the
+pipeline. That is, if the pipeline will only produce any output after
```

Comment:
This might read better as a parenthetical ("[...] in the pipeline (that is, if [...] supplied).").